### PR TITLE
Add attention mask for decoder whisper model

### PR DESCRIPTION
**Pull request by eshiryae**

### CODE REVIEW

**Review by dmatveev — APPROVED** [review]
LGTM

File: src/cpp/src/whisper/pipeline_static.cpp [hunk 1]
```diff
@@ -117,21 +119,27 @@ void update_past_key_value(ov::InferRequest& source, ov::InferRequest& dest, con
...
+    std::copy_n(
+        &init_ids[0], init_ids.size(),
+        padded_input_ids.data<int64_t>()
+    );
+
+    // NB: attention_mask format: [1, 1, 0, 0]
+    auto padded_attention_mask = decoder.get_tensor("attention_mask");
+    auto* padded_mask_data = padded_attention_mask.data<int64_t>();
+    std::fill_n(padded_mask_data, padded_attention_mask.get_size(), 0u);
+    std::fill_n(padded_mask_data, init_ids.size(), 1u);
```
Comment (AlexanderKalistratov) [inline src/cpp/src/whisper/pipeline_static.cpp:144]: Shouldn't we also add an assert for `padded_attention_mask.get_size()`:
```C++
OPENVINO_ASSERT(padded_attention_mask.get_size() >= init_ids.size());
```
?

File: src/cpp/src/whisper/pipeline_static.cpp [hunk 2]
```diff
@@ -170,8 +178,7 @@ ov::Tensor decode_with_past(ov::InferRequest& decoder_with_past,
                             ov::genai::RawPerfMetrics& raw_metrics) {
     decoder_with_past.get_tensor("input_ids").data<int64_t>()[0] = input_id;
     decoder_with_past.get_tensor("cache_position").data<int64_t>()[0] = position_id;
-    // FIXME: Is "attention_mask" supposed to be f16?
-    decoder_with_past.get_tensor("attention_mask").data<ov::float16>()[position_id - 1] = 0u;
+    decoder_with_past.get_tensor("attention_mask").data<float>()[position_id - 1] = 0u;
```
Comment (AlexanderKalistratov) [inline src/cpp/src/whisper/pipeline_static.cpp]: For `position_id == 0` we would get index `-1`. Is it ok?

File: src/cpp/src/whisper/pipeline_static.cpp [hunk 3]
```diff
@@ -393,20 +398,130 @@ void add_attention_mask_input(std::shared_ptr<ov::Model> model) {
...
+            if (transform_cross_attn) {
+                // Cross attn
+                OPENVINO_ASSERT(!cross_attn_nodes.empty());
+                // FIXME: Should be taken from topology - don't hardcode!!!
+                auto shape_cst = std::make_shared<v0::Constant>(
+                    ov::element::i64,
+                    ov::Shape{2},
+                    std::vector<float>{MAX_PROMPT_LEN, 1}
+                );
+                // FIXME: Avoid "1500" hardcode!!!
```
Comment (smirnov-alexey) [inline src/cpp/src/whisper/pipeline_static.cpp]: Is it going to be resolved in this PR or is it a non-blocker?

File: src/cpp/src/whisper/pipeline_static.cpp [hunk 4]
```diff
@@ -393,20 +398,130 @@ void add_attention_mask_input(std::shared_ptr<ov::Model> model) {
...
+            );
+
+            for (auto self_attn : self_attn_nodes) {
+                self_attn->input(3).replace_source_output(select->output(0));
+            }
+
+            if (transform_cross_attn) {
+                // Cross attn
+                OPENVINO_ASSERT(!cross_attn_nodes.empty());
+                // FIXME: Should be taken from topology - don't hardcode!!!
```
Comment (smirnov-alexey) [inline src/cpp/src/whisper/pipeline_static.cpp]: Is it going to be resolved in this PR or is it a non-blocker?

File: src/cpp/src/whisper/pipeline_static.cpp [hunk 5]
```diff
@@ -393,20 +398,130 @@ void add_attention_mask_input(std::shared_ptr<ov::Model> model) {
...
+                    ov::Shape{2},
+                    std::vector<float>{MAX_PROMPT_LEN, 1}
+                );
+                // FIXME: Avoid "1500" hardcode!!!
+                auto target_shape = std::make_shared<v0::Constant>(
+                    ov::element::i64,
+                    ov::Shape{2},
+                    std::vector<float>{MAX_PROMPT_LEN, 1500}
+                );
+                // FIXME: Must be transpose if batch present
```
Comment (smirnov-alexey) [inline src/cpp/src/whisper/pipeline_static.cpp:480]: Is it going to be resolved in this PR or is it a non-blocker?

File: src/cpp/src/whisper/pipeline_static.cpp [hunk 6]
```diff
@@ -117,21 +119,27 @@ void update_past_key_value(ov::InferRequest& source, ov::InferRequest& dest, con
...
-    const size_t seq_length = input_ids_tensor.get_shape()[1];
-
-    OPENVINO_ASSERT(seq_length >= init_ids.size());
-
-    auto input_ids_data = input_ids_tensor.data<int64_t>();
-    std::copy(init_ids.begin(), init_ids.end(), input_ids_data);
+    // NB: input_ids format: [token1, token2, pad, pad]
+    auto padded_input_ids = decoder.get_tensor("input_ids");
+    OPENVINO_ASSERT(padded_input_ids.get_size() >= init_ids.size());
+    std::fill_n(padded_input_ids.data<int64_t>(), padded_input_ids.get_size(), 0u);
```
Comment (esmirno) [inline src/cpp/src/whisper/pipeline_static.cpp:133]: potential issue if input_ids has non int64 type, since .data<int64_t>() might cast anyway - there is no typecheck of storage elements, and possible memory overflow could occur - suggest adding either type check after get_tensor("input_ids") or direct check for get_byte_size() or other mechanics

File: src/cpp/src/whisper/pipeline_static.cpp [hunk 7]
```diff
@@ -189,14 +196,14 @@ void zero_past_key_values(ov::InferRequest& request) {
...
-    auto attention_mask = decoder_with_past.get_tensor("attention_mask");
-    auto* attention_mask_ptr = attention_mask.data<ov::float16>();
-    std::fill(attention_mask_ptr, attention_mask_ptr + init_ids_size, 0);
-    std::fill(attention_mask_ptr + init_ids_size, attention_mask_ptr + attention_mask.get_size() - 2, 1);
-    attention_mask_ptr[attention_mask.get_size() - 2] = 0;
-    attention_mask_ptr[attention_mask.get_size() - 1] = 1;
+    // NB: Prepare attetion mask to be in a format [0, 0, 0, 1, 1, 1, ..., 1, 0, 1]
+    auto padded_attention_mask = decoder_with_past.get_tensor("attention_mask");
+    auto* padded_mask_data = padded_attention_mask.data<float>();
+    std::fill(padded_mask_data, padded_mask_data + init_ids_size, 0);
```
Comment (AlexanderKalistratov) [inline src/cpp/src/whisper/pipeline_static.cpp:207]: Shouldn't we use the same approach in `set_decoder_input_ids` https://github.com/openvinotoolkit/openvino.genai/blob/183de4cfdcb65253611dd462b040feabe739a76c/src/cpp/src/whisper/pipeline_static.cpp#L141

```C++
    const auto padded_mask_size = padded_attention_mask.get_size();
    std::fill(padded_mask_data, &padded_mask_data[init_ids.size()], 1u);
    std::fill(&padded_mask_data[init_ids.size()], &padded_mask_data[padded_mask_size], 0u);

```

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 2 above)
Comment (esmirno) [reply src/cpp/src/whisper/pipeline_static.cpp]: it is ok while =0u, might be writing to unexpected memory location

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 7 above)
Comment (AlexanderKalistratov) [inline src/cpp/src/whisper/pipeline_static.cpp:207]: Shouldn't we also add some assert checks on sizes?
```C++
OPENVINO_ASSERT(padded_attention_mask.get_size() >= init_ids_size);
OPENVINO_ASSERT(padded_attention_mask.get_size() >= 2);
```

File: src/cpp/src/whisper/pipeline_static.cpp [hunk 8]
```diff
@@ -393,20 +398,130 @@ void add_attention_mask_input(std::shared_ptr<ov::Model> model) {
...
+    using namespace ov::pass::pattern;
+    using namespace ov::op;
+    class AttentionMaskInput : public ov::pass::MatcherPass {
+    public:
+        OPENVINO_MATCHER_PASS_RTTI("AttentionMaskInput");
+
+        AttentionMaskInput(std::shared_ptr<ov::Model> model, bool transform_cross_attn) {
+            std::vector<std::shared_ptr<ov::Node>> self_attn_nodes;
+            std::vector<std::shared_ptr<ov::Node>> cross_attn_nodes;
+            for (auto node : model->get_ops()) {
```
Comment (AlexanderKalistratov) [inline src/cpp/src/whisper/pipeline_static.cpp]:
```C++
for (auto&& node : model->get_ops()) {
```

File: src/cpp/src/whisper/pipeline_static.cpp [hunk 9]
```diff
@@ -393,20 +398,130 @@ void add_attention_mask_input(std::shared_ptr<ov::Model> model) {
...
+                ov::Shape{1},
+                std::vector<float>{0}
+            );
+
+            auto equal = std::make_shared<v1::Equal>(mtpl->output(0), cst_1->output(0));
+            auto select = std::make_shared<v1::Select>(
+                equal->output(0), cst_0->output(0), cst_ninf->output(0)
+            );
+
+            for (auto self_attn : self_attn_nodes) {
```
Comment (AlexanderKalistratov) [inline src/cpp/src/whisper/pipeline_static.cpp]:
```C++
for (auto&& self_attn : self_attn_nodes) {
```

File: src/cpp/src/whisper/pipeline_static.cpp [hunk 10]
```diff
@@ -393,20 +398,130 @@ void add_attention_mask_input(std::shared_ptr<ov::Model> model) {
...
+                // FIXME: Must be transpose if batch present
+                auto reshape = std::make_shared<v1::Reshape>(cvt->output(0), shape_cst->output(0), false);
+                auto equal = std::make_shared<v1::Equal>(reshape->output(0), cst_1->output(0));
+                auto select = std::make_shared<v1::Select>(
+                    equal->output(0), cst_0->output(0), cst_ninf->output(0)
+                );
+                auto broadcast = std::make_shared<v3::Broadcast>(select->output(0), target_shape->output(0));
+                auto unsq1 = std::make_shared<v0::Unsqueeze>(broadcast->output(0), cst_0->output(0));
+                auto unsq2 = std::make_shared<v0::Unsqueeze>(unsq1->output(0), cst_1->output(0));
+                for (auto cross_attn_node : cross_attn_nodes) {
```
Comment (AlexanderKalistratov) [inline src/cpp/src/whisper/pipeline_static.cpp]:
```C++
for (auto&& cross_attn_node : cross_attn_nodes) {
```

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 2 above)
Comment (eshiryae) [reply src/cpp/src/whisper/pipeline_static.cpp]: We'll never get here `-1` as it's decoder_with_past run and `position_id` passed to this function at least 1.

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 6 above)
Comment (eshiryae) [reply src/cpp/src/whisper/pipeline_static.cpp:133]: Added assert for element type

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 1 above)
Comment (eshiryae) [reply src/cpp/src/whisper/pipeline_static.cpp:144]: Added

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 7 above)
Comment (eshiryae) [reply src/cpp/src/whisper/pipeline_static.cpp:207]: Added

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 5 above)
Comment (eshiryae) [reply src/cpp/src/whisper/pipeline_static.cpp:480]: It's not a blocker as NPU doesn't support now batch size >1, there's a check for this and resize in function full_decode()

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 3 above)
Comment (eshiryae) [reply src/cpp/src/whisper/pipeline_static.cpp]: Fixed

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 4 above)
Comment (eshiryae) [reply src/cpp/src/whisper/pipeline_static.cpp]: Fixed

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 8 above)
Comment (eshiryae) [reply src/cpp/src/whisper/pipeline_static.cpp]: Fixed

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 9 above)
Comment (eshiryae) [reply src/cpp/src/whisper/pipeline_static.cpp]: Fixed

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 10 above)
Comment (eshiryae) [reply src/cpp/src/whisper/pipeline_static.cpp]: Fixed

File: src/cpp/src/whisper/pipeline_static.cpp (hunk 7 above)
Comment (eshiryae) [reply src/cpp/src/whisper/pipeline_static.cpp:207]: Why not `std::fill()`?
I suppose it's more understandable this way how we're filling attention mask here

File: src/cpp/src/whisper/pipeline_static.cpp [hunk 11]
```diff
@@ -815,6 +915,7 @@ std::shared_ptr<ov::Model> prepare_decoder_with_past_model(std::shared_ptr<ov::M
     expose_runtime_states_as_inputs(decoder_with_past_model);
 
     decoder_with_past_model->reshape({{"input_ids", ov::PartialShape({-1, 1})}});
+    decoder_with_past_model->set_friendly_name("Model6");
```
Comment (dmatveev) [inline src/cpp/src/whisper/pipeline_static.cpp:918]: I believe modes get their names automatically, what's the reason to rename this particular one?

Comment (eshiryae) [reply src/cpp/src/whisper/pipeline_static.cpp:918]: As we're cloning model they will have the same name.
Just aligned names here with model names when we have 3 models.

Comment (dmatveev) [reply src/cpp/src/whisper/pipeline_static.cpp:918]: I'd give them some semantics then
